src/core/request/deltafetch.js

```diff
@@ -46,132 +42,68 @@ export class DeltaFetchRequest extends KinveyRequest {
-              const response = reduce(responses, (result, response) => {
-                if (response.isSuccess()) {
-                  const headers = result.headers;
-                  headers.addAll(response.headers);
-                  result.headers = headers;
-                  result.data = result.data.concat(response.data);
-                }
-
-                return result;
-              }, new Response({
-                statusCode: StatusCode.Ok,
-                data: []
-              }));
-
-              response.data = response.data.concat(values(cacheDocuments));
-
-              if (this.query) {
-                const query = new Query(result(this.query, 'toJSON', this.query));
-                query.skip = 0;
-                query.limit = 0;
-                response.data = query.process(response.data);
-              }
-
-              return response;
+        return repo.read(QUERY_CACHE_COLLECTION_NAME, queryCacheQuery)
+          .then((queries) => {
+            const request = new KinveyRequest({
+              method: RequestMethod.GET,
+              url: this.url,
+              headers: this.headers,
+              authType: this.authType,
+              query: this.query,
+              timeout: this.timeout,
+              client: this.client,
+              properties: this.properties,
+              skipBL: this.skipBL,
+              trace: this.trace,
+              followRedirect: this.followRedirect,
+              cache: this.cache
             });
-        }
 
-        const request = new KinveyRequest({
-          method: RequestMethod.GET,
-          url: this.url,
-          headers: this.headers,
-          authType: this.authType,
-          query: this.query,
-          timeout: this.timeout,
-          client: this.client,
-          properties: this.properties,
-          skipBL: this.skipBL,
-          trace: this.trace,
-          followRedirect: this.followRedirect,
-          cache: this.cache
-        });
-        return request.execute();
+            if (isArray(queries) && queries.length > 0) {
+              const since = queries[0].lastRequest;
+              request.url = urljoin(this.url, url.format({
+                pathname: '_deltaset',
+                query: { since }
+              }));
+              return request.execute();
+            }
+
+            return request.execute();
+          })
+          .then((response) => {
+            // Makse sure the response from DeltaSet is normalized
+            const { data } = response;
+
+            if (!Object.prototype.hasOwnProperty.call(data, 'changed')) {
+              data.changed = data;
+            }
+
+            if (!Object.prototype.hasOwnProperty.call(data, 'deleted')) {
+              data.deleted = [];
+            }
+
+            response.data = data;
+            return response;
+          })
+          .then((response) => {
+            return repo.create(QUERY_CACHE_COLLECTION_NAME, {
```

Comment:
Shouldn't this be an `update()`? Which is actually an upsert, which is what we are trying to do, no?

Reply:
Yes it should be an update.